### ISSUE TITLE
Vote: Prefix parameters.required with $ to indicate it is reserved

### ISSUE THREAD

**Issue by carolynvs** [deislabs/cnab-spec#178]
In our community meeting today, it was brought up that the recent json schema changes have caused us to inadvertently cause the parameter name "required" to be an undocumented reserved word and it is not possible for someone to use that as a parameter name. 

A suggestion was to prefix required with `$` and make it reserved, similar to how we have already done so for `$comment`.

So instead of 

```json
"parameters": {
    "required": [ "param1", "param2" ]
}
```

it would become

```json
"parameters": {
    "$required": [ "param1", "param2" ]
}
```

As notetaker for the meeting I was asked to bring this to a vote in an issue.  Please vote on this using the 👍 and 👎 reactions on this root issue. If you have an alternate solution or additional feedback, leave a reply!

**Comment by simongdavies** [comment]
Isn't it documented by the fact that it is in the spec and enforced by the schema?

**Comment by carolynvs** [comment]
It wasn't clear at all that we intended to camp on the name "required". In the PRs we talked about following json schema, but then in the implementation we didn't follow it 100% and ended up with this, where required is at a hierarchy level different from how json schema does it so that "required" is now not able to be used as a parameter. 

Being in the json isn't enough, and it's not that hard to add a line in the doc in plain english so that people don't ask the questions that at least 3 people asked the other week which sparked this issue. 😁

The ask at the time was to either fix the confusion somehow or at least write down that we are aware that we've made "required" a reserved word. In the meeting it was suggested that prefixing with $ was a good compromise, hence this issue. I'd love it if people would vote though... If this gets no traction at all I'll just submit some text explaining that it's essentially reserved and leave it at that.

**Comment by ryanmoran** [comment]
As an alternative, we could introduce a new field under parameters, maybe something like `fields` that would house the actual parameter definitions and then hoist the `required` field up to that same level:

```json
{
  "parameters": {
    "fields": {
      "some-param": {
        "type": "string",
        "destination": {
          "env": "SOME_PARAM"
        }
      }
    },
    "required": ["some-param"]
  }
}
```

After the addition of `outputs` I'd also been thinking a bit more about how we could have done this JSONSchema addition a bit better. For example, it may have made more sense to have separated the JSONSchema definitions from the parameter/credential/output specifications like this:

```json
{
  "definitions": {
    "some-definition": {
      "type": "string"
    }
  },
  "credentials": {
    "fields": {
      "some-credential": {
        "definition": "some-definition",
        "destination": {
          "env": "SECRET"
        }
      }
    }
  },
  "parameters": {
    "fields": {
      "some-parameter": {
        "definition": "some-definition",
        "destination": {
          "env": "PARAM"
        }
      }
    },
    "required": ["some-parameter"]
  },
  "outputs": {
    "fields": {
      "some-output": {
        "definition": "some-definition",
        "path": "/some/path"
      }
    }
  }
}
```

I think this leads to a slightly cleaner separation of concerns and some amount of possible reuse with these JSONSchema definitions that tend to get quite long and verbose. I also think that it makes sense for this to be something we would hold back to address until a v2.

**Comment by jeremyrickard** [comment]
@ryanmoran I initially added this to the 1.0 Milestone. I feel like we should resolve the issue about parameters and required before we declare the spec 1.0 since it's a breaking change. You mentioned holding back to a v2...if we make this change before 1.0 would that derail you?

**Comment by ryanmoran** [comment]
@jeremyrickard it would not block us. I was most trying not to block a path to 1.0. I agree it would be important to solve, and I’d prefer to see it sooner rather than later, just stating our flexibility with it sliding to later if it really needs to go.

**Comment by jeremyrickard** [comment]
@ryanmoran I'm in favor of your suggestion. We didn't seem to get any voting here so I'll ping the Slack channel to encourage. 

I think maybe your suggestion is two different issues, just for the sake of reviewing.

**Comment by technosophos** [comment]
I am in favor of @ryanmoran 's recommendation.

**Comment by jeremyrickard** [comment]
This is closed by #191 and #192.
